SCons/Tool/msvc.xml

```diff
@@ -578,5 +579,65 @@ and also before &f-link-env-Tool; is called to ininitialize any of those tools:
 </summary>
 </cvar>
 
+<cvar name="MSVC_NOTFOUND_POLICY">
+<summary>
+<para>
+Specify the scons behavior when the Microsoft Visual C/C++ compiler is not detected.
+</para>
+
+<para>
+    The <envar>MSVC_NOTFOUND_POLICY</envar> specifies the &scons; behavior when no msvc versions are detected or
+    when the requested msvc version is not detected.
+</para>
+
+<para> 
+The valid values for <envar>MSVC_NOTFOUND_POLICY</envar> and the corresponding &scons; behavior are:
+</para>
+
+<variablelist>
+
+<varlistentry>
+<term><parameter>'Error' or 'Exception'</parameter></term>
+<listitem>
+<para>
+Raise an exception when no msvc versions are detected or when the requested msvc version is not detected.
```

Comment:
Hmmm.  Need to think about it.  I may not be interepreting the question correctly.

The thinking was that unless the exception is caught and ignored by the user, the build is going to bail out due to the exception.  The other two are most likely going to fail later if the environment is attached to a program/dll as the dreaded cl not found message will probably arise.

The exception should only be raised when the user is requesting an msvc build.  If there are non-msvc build tools and the user does not explicitly set MSVC_VERSION or list one of the msvc tools, there should be no action taken as the non-msvc tool should be used.  Only when a non-default tools list is specified that includes one or more of the msvc tools and/or MSVC_VERSION is specified will the policy come into play.

In retrospect, it probably would have made sense to have a hook in the platform code that notifies the msvc code that there are no tools detected and msvc is being forced as the default tool.

If this was not even close to the answer to the question you're asking, I'll try again.

Reply:
> Sorry to be nitpicking. The middle paragraph is a good explantion. But this wording doesn't make that clear:

>>Raise an exception when no msvc versions are detected

No apologies necessary.  If it isn't clear, it isn't clear.

> That could be read as meaning if I have mingw installed, and I somehow foolishly set this flag to Exception, then it's going to bail if there's no msvc. You say it won't unless I also explicitly used MSVC_VERSION. I think we need to reflect that somehow.

You're right.  The intent is that the policy is only applied if there is a non-default msvc tool request or MSVC_VERSION request.

> A bit of the same applies to the warn policy case: does it warn if there's no msvc, but there's another valid compiler and I didn't ask for msvc?

No.  Again you're right.

The intention of the PR is to suppress the warning/error (i.e., the policy is ignored) when there are no versions of msvc detected and when: using the default tools list, a non-default tool list that does not contain any of the msvc tools, and no MSVC_VERSION specification.

> For that matter - is it only MSV?_VERSION that triggers this? What happens if I use tools=[msvc, mslink]? Is that detected by the code as somehow different from SCons going through its default list looking for msvc?

It is MSV?_VERSION or a non-default tool list containing msvc tools (msvc, mslink, mslib).

The comments in the SConstruct test cases in https://github.com/SCons/scons/pull/4132#issuecomment-1145085003 illustrate this example and a few others.

Reply:
```
...
The MSVC_NOTFOUND_POLICY specifies the scons behavior when no msvc versions are detected or
when the requested msvc version is not detected.

The MSVC_NOTFOUND_POLICY is applied only when MSVC_VERSION is specified and/or when a non-default
tool list is specified that includes one or more of the msvc tools (e.g., msvc, mslink, mslib, etc.).
...
```

Any clearer?

Reply:
In the PDF, the bullet items seem to be left AND right justified which results in some bizarre spacing.  Is this normal?  Can it disabled inline?

Reply:
Approximate rendering, changed "enforced if" to "applied when", changed "ignored if" to "ignored when", left the six bullets alone (novelty has worn off).

```
MSVC_NOTFOUND_POLICY
    Specify the scons behavior when the Microsoft Visual C/C++ compiler is not detected.

    The MSVC_NOTFOUND_POLICY specifies the scons behavior when no msvc versions are detected or 
    when the requested msvc version is not detected.

    The valid values for MSVC_NOTFOUND_POLICY and the corresponding scons behavior are:

    'Error' or 'Exception'
         Raise an exception when no msvc versions are detected or when the requested msvc version is
         not detected.

    'Warning' or 'Warn'
        Issue a warning and continue when no msvc versions are detected or when the requested msvc version is
        not detected. Depending on usage, this could result in build failure(s).

    'Ignore' or 'Suppress'
        Take no action and continue when no msvc versions are detected or when the requested msvc version is
        not detected. Depending on usage, this could result in build failure(s).

    Note: in addition to the camel case values shown above, lower case and upper case values are accepted
    as well.

    The MSVC_NOTFOUND_POLICY is applied when any of the following conditions are satisfied:
    * $MSVC_VERSION is specified, the default tools list is implicitly defined (i.e., the tools list is 
      not specified), and the default tools list contains one or more of the msvc tools.
    * $MSVC_VERSION is specified, the default tools list is explicitly specified (e.g., tools=['default']), 
      and the default tools list contains one or more of the msvc tools.
    * A non-default tools list is specified that contains one or more of the msvc tools (e.g., 
      tools=['msvc', 'mslink']).

    The MSVC_NOTFOUND_POLICY is ignored when any of the following conditions are satisfied:
    * $MSVC_VERSION is not specified and the default tools list is implicitly defined (i.e., the tools 
      list is not specified).
    * $MSVC_VERSION is not specified and the default tools list is explicitly specified (e.g., 
      tools=['default']).
    * A non-default tool list is specified that does not contain any of the msvc tools (e.g., 
      tools=['mingw']).

    When MSVC_NOTFOUND_POLICY is not specified, the default scons behavior is to issue a warning and continue
    subject to the conditions listed above. The default scons behavior may change in the future.
```
Some of the PDF rendering is funky.  Will push shortly...

Reply:
> In the PDF, the bullet items seem to be left AND right justified which results in some bizarre spacing. Is this normal? Can it disabled inline?

Haven't looked at the PDF versions for a long time.  Bill may disagree, but I consider them unimportant. :)  Don't want to send you down a rathole, so Do Not Go Look Here, but just for info purposes, the things that are controllable for the pdf output are in a stylesheet named `pdf.xsl` - there's a separate one for each of the four original docbook documents, only two of which are still maintained - man and user.